### ISSUE TITLE
create a new apis equivalent to the aggregate api but with different output format

### ISSUE THREAD

**Issue by hongsudt** [informatics-isi-edu/ermrest#98]

**Comment by karlcz** [comment]
@hongsudt @robes @RFSH @shinyichen 

I think this revised issue description is the sort of ermrest api extension we need for efficiently retrieving foreign entity data alongside the main entity.  It would work for both record2 and recordset apps.

Does this seem right from your understanding too?

**Comment by karlcz** [comment]
I revised the issue description again.  This needs more thought and design work as it is precariously close to the previously deferred goal of having scalar subqueries in the projection list...

**Comment by karlcz** [comment]
The complicating factors we need to consider in the design:

1. The messy  `(...)=(...)` link syntax for the general case
2. The desired projection behavior which does not affect the cardinality of the result set, i.e. it is not adding joins to the main entity path implicitly or anything like that
3. Which generalizations do we include or exclude (and can they be implemented in phases while maintaining a coherent API design):
  - Outbound references ( _path cols_ ) -> ( _schema cols_ )
  - Inbound references ( _path cols_ ) <- ( _schema cols_ )
  - Arbitrary joins ( _path cols_ ) == ( _schema cols_ ) not involving foreign key constraints
  - Multiple steps, i.e. to navigate across association tables
  - Filtering of projected entity set
  - Sub-projection to control how each projected entity is represented
  - Aggregation to control how projected entity sets are reduced to an output representation
4. The projection result structures/encodings. Are they chosen by different projection operators or by the ermrest query planner?
  - `null` due to NULL outbound reference?
  - `{ ... }` due to non-NULL outbound reference?
  - `[]` or `[ { ... }, ... ]` due to inbound references, arbitrary joins, multiple steps or require separate array projection operator?

**Comment by karlcz** [comment]
@hongsudt @RFSH @robes @shinyichen @mikedarcy 

This general sub-query feature is blocked on design and invasive code refactoring.  In the interim, outer joins in #99 may be sufficient for some simpler use cases where you just want a core entity and several optional extensions that it points at by nullable foreign key.

`/attributegroup/A:=T1/B:=left(A:fk1)=(T2:id)/C:=left(A:fk2)=(T3:id)/$A/id;a:=array(A:*),b:=array(B:*),c:=array(C:*)`

This will produce an output very similar to that in the issue description, only marred by the redundant row key and the wrapping of each sub-object as an array of length 1:

```
[
  { 
    "id": X,
    "a": [{ columns of T1, ... }], 
    "b": [{ columns of T2 entity referenced by T1:fk1, ... }], 
    "c": [{ columns of T3 entity referenced by T1:fk2, ... }]
  },
  ...
]
```

Furthermore, you need to understand your model enough to include the main entity key column (or columns) as the output group key. The use of the `left` keyword to enable left outer joins means that this will work even if the foreign key values happen to be NULL.  With an inner join, entities in T1 with NULL outbound foreign keys would be excluded.

**Comment by karlcz** [comment]
A few more UX discussions have exposed more feature desires for the generalized sub-query idea. GUI app writers would like more control over the aggregate reduction mechanism than is currently provided by the `/attributegroup/` API.

1. Rather than always grouping and aggregating over a general cross-product, sometimes the apps want to combine the `/entity/` or `/attribute/` API style navigation with aggregation. They would like to expand a set of *reachable entities* and then reduce that set with aggregate functions. They do not want to reduce the cross-product, distorting the population with multipath duplication.
2. This entity-based aggregation should also be combined with grouping mechanisms to return a sub-population aggregate per parent record. This is often written in idiomatic SQL using correlated sub-queries.
3. Sometimes the apps want to use aggregate functions like `array` to return sets of values, but they would want to apply a limit and a customized sort order, e.g. to only fetch the *first page* from a potentially large result. If performing grouped aggregations, they want a limited "page" per grouped result row to efficiently retrieve an initial result set. (Further scrolling through the pages would throw away this outer grouped structure and drill into a single group's result using a different ermrest query form, e.g. as the user clicks a "more..." link.)
